### ISSUE TITLE
Apply discount factor to climate benefits as they occur (Lashof)

### ISSUE THREAD

**Issue by badgley** [carbonplan/ton-year#17]
# Expected Behavior
Discounting of the radiative forcing and/or temperature benefits of any delayed emissions should be credited _as those delays occur_. Schematically, we can think about delayed emissions looking something like this: 

![image](https://user-images.githubusercontent.com/1392400/145229842-effa52cc-ee86-4b98-a440-878efe4e966b.png)

In this figure, we have three scenarios: a scenario where we delay emission of 1 GtC, entirely avoid emission of 1 GtC, and a baseline scenario where there is no delay or avoid. The climate benefits of delay are represented by the difference between the blue `default` curve and the green `delay` curve. From an atmospheric standpoint, the benefits of delay accumulate throughout the duration of the delay period. 

# Actual Behavior
The current implementation of the `lashof` method only considers the climate benefits of delaying emission delay at the _end_ of the time horizon, as opposed to at the time of delay. I think this is physically incorrect, though I'm more than open to being wrong! 

Conceptually, `lashof` calculates the integral of the CO2 impulse response function  under i) a baseline scenario and ii) a delayed scenario. Schematically, the benefit is often depicted as "the area [of the delayed IRF] pushed beyond the time horizon" (Fearnside, Moura-Costa, and Lashof, 2000; FMCL). However, it's not immediately clear that this means the climate benefits of delay should only be evaluated at the end of the time horizon. Instead, I think the discussion of "pushed beyond the time horizon" is actually an attempt to _build intuition_ around the underlying math/calculus. By my reading, they're saying something like "you take the integral of the IRF over the time horizon; you then take the integral of the _shifted_ IRF over the time horizon; the difference in those areas is the climate benefit! If integrals give you any trouble, we can show this schematically -- the difference in those two integrals is _actually_ the same as the area of the shifted IRF curve that falls outside the time horizon." 

In short, I do not think that FMCL argue that climate benefits of delay accumulate at `time horizon + delay`.  Instead, I think there are strong _physical_ reasons we should calculate climate benefits at the time of delay. This becomes especially relevant in discounted scenarios. 

# Proposed solution
I think we just need to lightly edit the [lashof method](https://github.com/carbonplan/ton-year/blob/7ffc917a57106b2eb2325c110b7e252be2009b2f/tonyear/core.py#L157).  

Intuitively, it should just be something like 

```python
benefit = np.trapz(baseline) - np.trapz(scenario[:time_horizon])
``` 

as opposed to
```python
benefit = np.trapz(res['scenario'][time_horizon:])
```

When I do this, it ~works for longer time delays but there is up to 20 percent divergence at very short time delays: 

![image](https://user-images.githubusercontent.com/1392400/145239364-af5fda77-4c0a-486b-9789-3eb70ba6288d.png)

So maybe there is an off by one error and/or something about `numpy.trapz` that we need to sort out. But before really sinking my teeth in, I thought I would submit this to get feedback on whether or not we even think this is an issue!

**Comment by freyac** [comment]
I think you are describing what's currently implemented as the `ipcc method`(which, agreed, is of like a Lashof #2).

`    elif method == 'ipcc':
        # The IPCC method calculates calculates the ton-year benefit of an emission at t=delay
        # as the difference between the baseline atmospheric cost and the scenario atmospheric
        # cost, which is calculated over the period delay<=t<=time_horizon.
        scenario = np.concatenate((np.zeros(delay), baseline))[:time_horizon_timesteps]
        scenario = get_discounted_curve(discount_rate, scenario)
        benefit = baseline_atm_cost - np.trapz(scenario[delay:])`

I thought it was useful to implement both to be able to recreate all the figures / framings seen in literature, but I agree with you that the discounting interaction makes much more sense when applied to `ipcc`. 

Does that resolve your concern? Do you think we should do some re-naming of the methods?

**Comment by badgley** [comment]
Right, my concern is that `lashof` as currently implemented is based on a misreading (albeit an understandable misreading!) of how the approach was actually intended to work. My reading is that the whole "pushing beyond the time horizon" is a graphical/schematic/intuitive description. It's not clear to me that anyone in the literature is actually advocating climate benefits be calculated in the way we've implemented in `lashof`. 

If the above is true, I think we should definitely rename `ipcc` to `lashof` and link to either i) this issue or ii) some other description of this potential discrepancy.

**Comment by freyac** [comment]
Totally hear what you're saying.

I do think it's important to be able to clearly map the methods to how prominent sources have talked about the approaches. I'm thinking particularly of the [ipcc report's](https://archive.ipcc.ch/ipccreports/sres/land_use/index.php?idp=74), which doesn't name the difference-between-two-curves approach, and represents `lashof` as the stick-out-past-the-end approach, both textually and graphically. 

 I'd feel great about renaming: `ipcc` --> `lashof` and `lashof` --> `lashof_graphical` or something similar, with a clear explanation.

**Comment by badgley** [comment]
Definitely okay with something like that, but I still have some hesitations. 

Specifically, what about discounting? I do not think that `lashof_graphical` should ever be discounted. There is no example of discounting the benefit curve in the literature that I've come across, so we're sort of in new territory here. It just doesn't feel "right" to discount those climate benefits at `time horizon + delay`, given that those benefits occur at time `start + delay`. It's exactly this inconsistency that got me reading FMCL more closely and come to the conclusion that the whole 'pushed past the end of the curve' was mainly an intuitive description, as opposed to a methodological prescription.

**Comment by jhamman** [comment]
>I think we should clearly make the point about the two ways to visualize the Lashof benefit (and the fact that in some places they're represented as separate methods) in the main notebook and in our methods document.

Perhaps this can be a short _How-to_ in the documentation page.

**Comment by freyac** [comment]
Addressed by PR#21. Closing!
